### PR TITLE
Feature: WinDivert IPS functionality on Windows

### PR DESCRIPTION
**Pull request by smithjacobj**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/2455

Describe changes:
- Enables IPS functionality on Windows using the open-source (LGPLv3/GPLv2)
WinDivert 1.4.1 driver and API.

> From https://www.reqrypt.org/windivert-doc.html: "WinDivert is a
user-mode capture/sniffing/modification/blocking/re-injection package
for Windows Vista, Windows Server 2008, Windows 7, and Windows 8.
WinDivert can be used to implement user-mode packet filters, packet
sniffers, firewalls, NAT, VPNs, tunneling applications, etc., without
the need to write kernel-mode code."

- adds `--windivert [filter string]` and `--windivert-forward [filter
string]` command-line options to enable WinDivert IPS mode.
`--windivert[-forward] true` will open a filter for all traffic. See
https://www.reqrypt.org/windivert-doc.html#filter_language for more
information.

Limitation: currently limited to `autofp` runmode.

- Added interface offload query/disable/restore for Win32 via WMI (see `win32-syscall.h/c`)

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable): I do not currently have access to buildbot.

-----

This is a replacement PR for #3337

### CODE REVIEW

**Comment by victorjulien** [comment]
screenshot from 2018-04-14 10-55-10

This PR adds a new compile warning. This is w/o enabling windivert.

**Comment by victorjulien** [comment]
Could you clean up the commit message to look like a single commit instead of a squash of many?

File: configure.ac [hunk 1]
```diff
@@ -232,8 +232,8 @@
             RUST_LDADD="-ldl -lrt -lm"
             ;;
         *-*-mingw32*)
-            CFLAGS="${CFLAGS} -DOS_WIN32"
-            LDFLAGS="${LDFLAGS} -lws2_32"
+            CFLAGS="${CFLAGS} -DOS_WIN32 -DNTDDI_VERSION=NTDDI_VISTA -D_WIN32_WINNT=_WIN32_WINNT_VISTA"
```
Comment (victorjulien) [inline configure.ac]: should this be conditional based on whether windivert is enabled?

File: configure.ac [hunk 2]
```diff
@@ -1394,7 +1433,7 @@
 
         AC_ARG_WITH(libhtp_includes,
                 [  --with-libhtp-includes=DIR  libhtp include directory],
-                [with_libhtp_includes="$withval"],[with_libhtp_includes=no])
+                [with_libhtp_includes="$withval"],[with_libhtp_includes="no"])
```
Comment (victorjulien) [inline configure.ac]: Please make this a separate commit as it's unrelated to windivert

File: src/source-windivert.c [hunk 3]
```diff
@@ -0,0 +1,844 @@
...
+#endif
+
+#ifndef WINDIVERT
+/* Gracefully handle the case where no WinDivert support is compiled in */
+
+TmEcode NoWinDivertSupportExit(ThreadVars *, const void *, void **);
+
+void TmModuleReceiveWinDivertRegister(void)
+{
+    memset(&tmm_modules[TMM_RECEIVEWINDIVERT], 0, sizeof(TmModule));
```
Comment (victorjulien) [inline src/source-windivert.c]: instead of these individual memsets, I think we can memset the whole thing at startup? RegisterAllModules looks like a good candidate.

File: configure.ac (hunk 1 above)
Comment (smithjacobj) [reply configure.ac]: Good question. It _can_ be conditional, but the WMI offload query/disable code requires Vista or newer. I suspect you still want XP compatibility if possible? I can probably flag the functionality in an `#ifdef` on `NTDDI_VERSION`/`_WIN32_WINNT`, and have `--enable-windivert` switch those on by default.

File: src/source-windivert.c (hunk 3 above)
Comment (smithjacobj) [reply src/source-windivert.c]: Is this a request that I change this for the PR or a note for your reference? Right now, this seems to be the default behavior for other thread modules to zero themselves.

File: configure.ac (hunk 2 above)
Comment (smithjacobj) [reply configure.ac]: OK.

**Comment by smithjacobj** [comment]
> This PR adds a new compile warning. This is w/o enabling windivert.

Are you hoping to eliminate this warning in this PR? I can work on it, the changes to defining API versions mean that Windows provides its own definitions for POSIX host/network byte order conversions.

File: configure.ac (hunk 1 above)
Comment (victorjulien) [reply configure.ac]: I don't specifically care about XP. Whats the lowest OS version still supported by Microsoft for the general public?

File: src/source-windivert.c (hunk 3 above)
Comment (victorjulien) [reply src/source-windivert.c]: Hmm maybe I need more coffee, but I don't see the other modules doing this. I think it would be a cleaner solution to do the whole thing at once, but it's not a blocker. So if you're up for it, go ahead :)

**Comment by victorjulien** [comment]
In general PRs should not introduce new warnings. Perhaps it would make sense to do a separate commit / PR for the changing Windows API, then build windivert commit(s) on top of that?

**Comment by TechnikEmpire** [comment]
Just FYI, this warning coming from ws2tcpip.h is because mingw is garbage and most of their windows headers are wrong. The actual official header has the proper function definitions here with the const qualifier.

File: configure.ac (hunk 1 above)
Comment (smithjacobj) [reply configure.ac]: Windows 7 is the oldest supported now. Windows Vista support technically ended a year ago.

I made the change to make this conditional testing another issue, so it'll be optional without WinDivert.

**Comment by victorjulien** [comment]
On 24-04-18 16:00, Jacob Masen-Smith wrote:
> *@smithjacobj* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In configure.ac
> <https://github.com/OISF/suricata/pull/3339#discussion_r183740898>:
> 
>> @@ -232,8 +232,8 @@
>              RUST_LDADD="-ldl -lrt -lm"
>              ;;
>          *-*-mingw32*)
> -            CFLAGS="${CFLAGS} -DOS_WIN32"
> -            LDFLAGS="${LDFLAGS} -lws2_32"
> +            CFLAGS="${CFLAGS} -DOS_WIN32 -DNTDDI_VERSION=NTDDI_VISTA -D_WIN32_WINNT=_WIN32_WINNT_VISTA"
> 
> Windows 7 is the oldest supported now. Windows Vista support technically
> ended a year ago.
> 
> I made the change to make this conditional testing another issue, so
> it'll be optional without WinDivert.

Cool. Btw, I'm not opposed to changes breaking things for Vista, XP or
even lower.

File: src/source-windivert.c (hunk 3 above)
Comment (smithjacobj) [reply src/source-windivert.c]: Nope, I need more coffee, I was wrong. I based it on `source-nfq.c` but that uses a bunch of manual `NULL` assignments. I extrapolated the `memset` zeroing from that.

There currently doesn't seem to be any full zeroing of the `tmm_modules` allocation.

**Comment by smithjacobj** [comment]
It does look like I need to consider Win Server 2008 R2, which is end-of-life in 2020, but the major architectural changes happen around Vista and most of the API changes after that are mostly iterative, so it's not so hard maintaining Vista compatibility.
